flocker/acceptance/endtoend/test_dockerplugin.py

```diff
@@ -57,6 +78,46 @@ def run_python_container(self, cluster, address, docker_arguments, script,
         return cid
 
     @require_cluster(1)
+    def test_volume_persists_restart(self, cluster):
+        """
+        If a container with a volume is created with a restart policy of
+        "always", the container will restart with the same volume attached
+        after the Docker daemon is restarted.
+        """
+        # create a simple data HTTP python container, with the restart policy
+        data = random_name(self).encode("utf-8")
+        node = cluster.nodes[0]
+        http_port = 8080
+        host_port = find_free_port()[1]
+
+        volume_name = random_name(self)
+        self.run_python_container(
+            cluster, node.public_address,
+            {"host_config": create_host_config(
+                binds=["{}:/data".format(volume_name)],
+                port_bindings={http_port: host_port},
+                restart_policy={"Name": "always"}),
+             "ports": [http_port]},
+            SCRIPTS.child(b"datahttp.py"),
+            # This tells the script where it should store its data,
+            # and we want it to specifically use the volume:
+            [u"/data"])
+
+        # write some data to it via POST
+        d = post_http_server(self, node.public_address, host_port,
+                             {"data": data})
+        # assert the data has been written
+        d.addCallback(lambda _: assert_http_server(
+            self, node.public_address, host_port, expected_response=data))
+        # restart the Docker daemon
+        d.addCallback(lambda _: self.restart_docker(node.public_address))
+        # attempt to read the data back again; the container should've
+        # restarted automatically.
+        d.addCallback(lambda _: assert_http_server(
```

Comment:
There's a race condition here. The docker service might say that it is restarted before the container has fired up. The right thing to do is poll a few times with a timeout.